pkg/controller/logstorage/linseed/linseed_controller.go

```diff
@@ -0,0 +1,389 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package linseed
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	operatorv1 "github.com/tigera/operator/api/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/tigera/operator/pkg/common"
+	octrl "github.com/tigera/operator/pkg/controller"
+	"github.com/tigera/operator/pkg/controller/certificatemanager"
+	logstoragecommon "github.com/tigera/operator/pkg/controller/logstorage/common"
+	"github.com/tigera/operator/pkg/controller/options"
+	"github.com/tigera/operator/pkg/controller/status"
+	"github.com/tigera/operator/pkg/controller/utils"
+	"github.com/tigera/operator/pkg/controller/utils/imageset"
+	"github.com/tigera/operator/pkg/render"
+	relasticsearch "github.com/tigera/operator/pkg/render/common/elasticsearch"
+	"github.com/tigera/operator/pkg/render/logstorage/esgateway"
+	"github.com/tigera/operator/pkg/render/logstorage/linseed"
+	"github.com/tigera/operator/pkg/render/monitor"
+	"github.com/tigera/operator/pkg/tls/certificatemanagement"
+)
+
+var log = logf.Log.WithName("controller_logstorage_linseed")
+
+type LinseedSubController struct {
+	client         client.Client
+	scheme         *runtime.Scheme
+	status         status.StatusManager
+	clusterDomain  string
+	tierWatchReady *utils.ReadyFlag
+	usePSP         bool
+	multiTenant    bool
+}
+
+func Add(mgr manager.Manager, opts options.AddOptions) error {
+	if !opts.EnterpriseCRDExists {
+		return nil
+	}
+
+	// Create the reconciler
+	r := &LinseedSubController{
+		client:        mgr.GetClient(),
+		scheme:        mgr.GetScheme(),
+		clusterDomain: opts.ClusterDomain,
+		multiTenant:   opts.MultiTenant,
+		status:        status.New(mgr.GetClient(), "log-storage-access", opts.KubernetesVersion),
+	}
+	r.status.Run(opts.ShutdownContext)
+
+	// Create a controller using the reconciler and register it with the manager to receive reconcile calls.
+	c, err := controller.New("log-storage-access-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Determine how to handle watch events for cluster-scoped resources. For multi-tenant clusters,
+	// we should update all tenants whenever one changes. For single-tenatn clusters, we can just queue the object.
+	var eventHandler handler.EventHandler = &handler.EnqueueRequestForObject{}
+	if opts.MultiTenant {
+		eventHandler = utils.EnqueueAllTenants(mgr.GetClient())
```

Comment:
This PR introduces a new event handler type, that means "Trigger a reconcile for all tenants" - it's pretty convenient. I wonder what other simplifications we can make by developing some smarter EventHandler implementations... :thinking: